mbedtls/src/rng/hmac_drbg.rs

```diff
@@ -9,60 +9,79 @@
 use mbedtls_sys::types::raw_types::{c_int, c_uchar, c_void};
 use mbedtls_sys::types::size_t;
 pub use mbedtls_sys::HMAC_DRBG_RESEED_INTERVAL as RESEED_INTERVAL;
-use mbedtls_sys::{
-    hmac_drbg_random, hmac_drbg_reseed, hmac_drbg_seed, hmac_drbg_seed_buf,
-    hmac_drbg_set_prediction_resistance, hmac_drbg_update, HMAC_DRBG_PR_OFF, HMAC_DRBG_PR_ON,
-};
+use mbedtls_sys::*;
 
-use super::{EntropyCallback, RngCallback};
+use crate::rng::{EntropyCallback, RngCallback, RngCallbackMut};
 use crate::error::{IntoResult, Result};
 use crate::hash::MdInfo;
+use std::sync::Arc;
 
-define!(
-    #[c_ty(hmac_drbg_context)]
-    struct HmacDrbg<'entropy>;
-    const init: fn() -> Self = hmac_drbg_init;
-    const drop: fn(&mut Self) = hmac_drbg_free;
-);
+#[allow(dead_code)]
+pub struct HmacDrbg {
+    // Moving data causes dangling pointers: https://github.com/ARMmbed/mbedtls/issues/2147
+    // Storing data in heap and forcing rust move to only move the pointer (box) referencing it.
+    // The move will be faster. Access to data will be slower due to additional indirection.
+    inner: Box<hmac_drbg_context>,
+    entropy: Option<Arc<dyn EntropyCallback + 'static>>,
```

Comment:
I think we should allow the user to choose whether they use Arc or Box, for example like so:
```rust
use std::sync::Arc;

// somewhere in mbedtls public interface
pub trait IntoPinned: private::IntoPinnedPrivate {}
impl<T> IntoPinned for Box<T> {}
impl<T> IntoPinned for Arc<T> {}

// in mbedtls::private
mod private {
    use std::ops::Deref;
    use std::sync::Arc;

    enum PinnedStorage<T> {
        Box(Box<T>),
        Arc(Arc<T>),
    }

    impl<T> Deref for PinnedStorage<T> {
        type Target = T;
        
        fn deref(&self) -> &T {
            match self {
                PinnedStorage::Box(v) => &**v,
                PinnedStorage::Arc(v) => &**v,
            }
        }
    }

    pub trait IntoPinnedPrivate {
        type Contained;

        fn into_pinned(self) -> PinnedStorage<Self::Contained>;
    }

    impl<T> IntoPinnedPrivate for Box<T> {
        type Contained = T;

        fn into_pinned(self) -> PinnedStorage<T> {
            PinnedStorage::Box(self)
        }
    }

    impl<T> IntoPinnedPrivate for Arc<T> {
        type Contained = T;

        fn into_pinned(self) -> PinnedStorage<T> {
            PinnedStorage::Arc(self)
        }
    }
}
```